File: specter_exec/allowlist.py
"""Who may drive execution, and where it may run.

Both gates live here rather than in the caller, because this is the layer that
spawns a process as the host user. A caller that could be trusted to check would
not need the check.

Enforced identically whether the backend imports this directly (native) or the
host shim does on its behalf (Docker) — the same reason the module exists.
"""

from __future__ import annotations

import json
import os
import secrets
import sys
from pathlib import Path

# Where the token lives depends on how the backend was deployed, so resolution is
# ordered rather than fixed:
#   1. SPECTER_RUNNER_TOKEN_FILE  — explicit wins
#   2. /app/secrets/runner-token  — the mount a containerized backend can read;
#                                   ~/.specter is on the HOST and invisible to it
#   3. ~/.specter/runner-token    — the native default, and where the runner
#                                   writes when nothing overrides it
#
# The extraction of this module briefly lost this and kept only the runner's own
# path, which silently broke the containerized backend: it looked in /root and
# ignored the mounted file sitting beside it.
def _token_candidates() -> list[Path]:
    out = []
    override = os.environ.get("SPECTER_RUNNER_TOKEN_FILE")
    if override:
        out.append(Path(override))
    out.append(Path("/app/secrets/runner-token"))
    out.append(Path.home() / ".specter" / "runner-token")
    return out


# The path this process WRITES to. Readers use _token_candidates().
RUNNER_TOKEN_FILE = _token_candidates()[0]

# The approved-workspace list, synced from the backend. Overridable for the same
# reason as the token: a containerized backend cannot see the host's $HOME.
WORKSPACES_CONFIG = Path(
    os.environ.get("SPECTER_WORKSPACES_CONFIG", str(Path.home() / ".specter" / "workspaces.json"))
)

RUNNER_AUTH_HEADER = "X-Specter-Runner-Token"

# Reachable without a token: liveness and version only. They expose nothing and
# spawn nothing, and the backend needs /health before it holds a token.
UNAUTHENTICATED_PATHS = {"/health", "/version"}


def runner_token() -> str | None:
    """The shared secret, or None if the runner has not been provisioned."""
    for candidate in _token_candidates():
        try:
            token = candidate.read_text(encoding="utf-8").strip()
            if token:
                return token
        except OSError:
            continue
    return None


def ensure_runner_token() -> str:
    """Read the token, minting one on first start.

    Written 0600 -- it is the only thing standing between a local process and
    an agent running as this user.
    """
    existing = runner_token()
    if existing:
        return existing
    RUNNER_TOKEN_FILE.parent.mkdir(parents=True, exist_ok=True)
    token = secrets.token_urlsafe(32)
    RUNNER_TOKEN_FILE.write_text(token, encoding="utf-8")
    try:
        RUNNER_TOKEN_FILE.chmod(0o600)
    except OSError:
        pass
    return token


def approved_workspaces() -> list[Path] | None:
    """Approved roots, synced from the backend.

    Returns None when the file is absent or unreadable -- distinct from an empty
    list. Callers must FAIL CLOSED on None: a missing config means "not
    provisioned yet", never "allow everything".
    """
    try:
        raw = json.loads(WORKSPACES_CONFIG.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    paths = raw.get("paths") if isinstance(raw, dict) else raw
    if not isinstance(paths, list):
        return None
    out = []
    for entry in paths:
        if isinstance(entry, str) and entry.strip():
            out.append(Path(entry).expanduser().resolve())
    return out


def approved_workspace(path: str) -> tuple[Path | None, str]:
    """Resolve a requested workspace against the allowlist.

    Returns (resolved_path, "") when approved, or (None, reason) when not.

    Mirrors _approved_workspace_path in backend/app/routers/runs.py: a request is
    approved if it IS an approved root or sits inside one. Resolved first, so a
    symlink cannot point outside an approved tree and still match.
    """
    if not path or not str(path).strip():
        return None, "Workspace path is required."

    requested = Path(str(path)).expanduser().resolve()
    roots = approved_workspaces()

    if roots is None:
        return None, (
            "This runner has no approved-workspace list yet. Start the Specter "
            f"backend once to sync it, or write {WORKSPACES_CONFIG} yourself."
        )
    if not roots:
        return None, "No repositories are approved for agent execution."

    for root in roots:
        if requested == root or root in requested.parents:
            return requested, ""

    return None, f"Workspace path is not approved for agent execution: {requested}"



# ── self-check ───────────────────────────────────────────────────────────────
# Run:  python3 specter_exec/allowlist.py --self-check
#
# Every case here is a real failure. The extraction of this module out of the
# host runner dropped the multi-location token lookup and silently broke the
# containerized backend — it read /root/.specter and ignored the mounted file
# beside it. Nothing caught that, so these exist.

def _demo() -> None:
    import tempfile

    global WORKSPACES_CONFIG
    failed = 0

    def check(name: str, got, want) -> None:
        nonlocal failed
        ok = got == want
        print(f"  {'✓' if ok else '✗'} {name}" + ("" if ok else f"  got {got!r}, want {want!r}"))
        if not ok:
            failed += 1

    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp).resolve()
        (root / "repo" / "sub").mkdir(parents=True)
        (root / "other").mkdir()

        cfg = root / "workspaces.json"
        cfg.write_text(json.dumps({"paths": [str(root / "repo")]}))

        original = WORKSPACES_CONFIG
        WORKSPACES_CONFIG = cfg
        try:
            check("an approved root is allowed",
                  approved_workspace(str(root / "repo"))[0] is not None, True)
            check("a subdirectory of it is allowed",
                  approved_workspace(str(root / "repo" / "sub"))[0] is not None, True)
            check("an unapproved sibling is rejected",
                  approved_workspace(str(root / "other"))[0], None)
            # Resolved before comparison, or '..' walks straight out of the root.
            check("traversal out of an approved root is rejected",
                  approved_workspace(str(root / "repo" / ".." / "other"))[0], None)
            check("an empty path is rejected", approved_workspace("")[0], None)

            # Fail CLOSED. A missing config means "not provisioned", never
            # "allow everything" — the whole point of the gate.
            WORKSPACES_CONFIG = root / "does-not-exist.json"
            path, reason = approved_workspace(str(root / "repo"))
            check("a missing allowlist rejects rather than allows", path, None)
            check("and says what to do about it", "sync" in reason.lower(), True)

            # A corrupt file is also "not provisioned", not "allow everything".
            bad = root / "bad.json"
            bad.write_text("{not json")
            WORKSPACES_CONFIG = bad
            check("a corrupt allowlist fails closed",
                  approved_workspace(str(root / "repo"))[0], None)
        finally:
            WORKSPACES_CONFIG = original

        # THE REGRESSION: a containerized backend reads a mounted path, not $HOME.
        tok = root / "runner-token"
        tok.write_text("token-from-an-alternate-location")
        os.environ["SPECTER_RUNNER_TOKEN_FILE"] = str(tok)
        try:
            check("the token is found via SPECTER_RUNNER_TOKEN_FILE",
                  runner_token(), "token-from-an-alternate-location")
        finally:
            del os.environ["SPECTER_RUNNER_TOKEN_FILE"]

    print(f"\n  {'FAILED' if failed else 'all checks passed'}"
          + (f" — {failed} assertion(s)" if failed else ""))
    raise SystemExit(1 if failed else 0)


if __name__ == "__main__" and "--self-check" in sys.argv:
    _demo()

File: internal/exec/allowlist.go
// Package exec is Specter's execution engine: everything needed to run an agent
// on this machine, with nothing about how the request arrived.
//
// The CLI imports it directly. So does the server. That is the whole design —
// one implementation, two entry points, no HTTP between a caller and a
// subprocess it is about to spawn on the same machine.
package exec

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/navjyotnishant/specter-agent/internal/specterhome"
)

// Who may drive execution, and where it may run.
//
// Both gates live here rather than in the caller, because this is the layer that
// spawns a process as the host user. A caller that could be trusted to check
// would not need the check.

const (
	// AuthHeader carries the shared secret when execution is driven over HTTP.
	// Binding to localhost is not an authorization boundary: every local process
	// shares that address.
	AuthHeader = "X-Specter-Runner-Token"

	tokenEnv     = "SPECTER_RUNNER_TOKEN_FILE"
	allowlistEnv = "SPECTER_WORKSPACES_CONFIG"
)

// tokenCandidates is ordered, because the backend runs in two places and only
// one of them can see the host's home directory.
//
//  1. the explicit override
//  2. the mount a containerized backend can read
//  3. the native default, where the runner writes when nothing overrides it
//
// Losing step 2 during a refactor silently broke the containerized backend once
// already: it read /root/.specter and ignored the mounted file beside it.
func tokenCandidates() []string {
	// An explicit override is AUTHORITATIVE, not a first preference. Falling
	// through to a different token when the named file is missing is worse than
	// finding none: it silently uses a credential the operator did not point at,
	// and a test or a container would appear to work while reading the wrong
	// secret.
	if override := os.Getenv(tokenEnv); override != "" {
		return []string{override}
	}

	return []string{"/app/secrets/runner-token", specterhome.Path("runner-token")}
}

// RunnerToken returns the shared secret, or "" when unprovisioned.
func RunnerToken() string {
	for _, candidate := range tokenCandidates() {
		body, err := os.ReadFile(candidate)
		if err != nil {
			continue
		}
		if token := strings.TrimSpace(string(body)); token != "" {
			return token
		}
	}
	return ""
}

// EnsureRunnerToken reads the token, minting one on first use.
//
// Written 0600 and reused thereafter: regenerating on every start would
// invalidate whatever copy the backend already holds.
func EnsureRunnerToken() (string, error) {
	if existing := RunnerToken(); existing != "" {
		return existing, nil
	}

	path := tokenCandidates()[0]
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return "", fmt.Errorf("creating token directory: %w", err)
	}

	buf := make([]byte, 32)
	if _, err := rand.Read(buf); err != nil {
		return "", fmt.Errorf("generating token: %w", err)
	}
	token := base64.RawURLEncoding.EncodeToString(buf)

	if err := os.WriteFile(path, []byte(token), 0o600); err != nil {
		return "", fmt.Errorf("writing token: %w", err)
	}
	return token, nil
}

// AllowlistPath is where the approved-workspace list lives. Overridable for the
// same reason as the token: a containerized backend cannot see the host's home.
func AllowlistPath() string {
	if override := os.Getenv(allowlistEnv); override != "" {
		return override
	}
	return specterhome.Path("workspaces.json")
}

// approvedRoots reads the synced allowlist.
//
// A nil slice with no error is impossible here on purpose: the caller must be
// able to tell "no roots approved" from "the file is missing", because those
// demand opposite answers. Missing returns an error; empty returns an empty
// slice.
func approvedRoots(configPath string) ([]string, error) {
	body, err := os.ReadFile(configPath)
	if err != nil {
		return nil, fmt.Errorf("reading allowlist: %w", err)
	}

	var wrapper struct {
		Paths []string `json:"paths"`
	}
	if err := json.Unmarshal(body, &wrapper); err != nil {
		return nil, fmt.Errorf("parsing allowlist: %w", err)
	}

	out := make([]string, 0, len(wrapper.Paths))
	for _, raw := range wrapper.Paths {
		if strings.TrimSpace(raw) == "" {
			continue
		}
		resolved, err := filepath.EvalSymlinks(raw)
		if err != nil {
			// An approved path that no longer exists is not an error worth
			// failing the whole list over; it simply matches nothing.
			resolved = filepath.Clean(raw)
		}
		out = append(out, resolved)
	}
	return out, nil
}

// ApprovedWorkspace resolves a requested workspace against the allowlist.
//
// Returns the resolved path and "" when approved, or "" and a reason when not.
//
// Paths are resolved BEFORE comparison. Without that, "approved/../elsewhere"
// walks straight out of an approved root, and a symlink inside one points
// anywhere on the filesystem.
func ApprovedWorkspace(path, configPath string) (string, string) {
	if strings.TrimSpace(path) == "" {
		return "", "Workspace path is required."
	}

	requested, err := filepath.EvalSymlinks(path)
	if err != nil {
		requested = filepath.Clean(path)
	}
	requested, _ = filepath.Abs(requested)

	roots, err := approvedRoots(configPath)
	if err != nil {
		// FAIL CLOSED. Not provisioned is not permission.
		return "", fmt.Sprintf(
			"This runner has no approved-workspace list yet. Start the Specter "+
				"backend once to sync it, or write %s yourself.", configPath)
	}
	if len(roots) == 0 {
		return "", "No repositories are approved for agent execution."
	}

	for _, root := range roots {
		if requested == root || strings.HasPrefix(requested, root+string(filepath.Separator)) {
			return requested, ""
		}
	}
	return "", fmt.Sprintf("Workspace path is not approved for agent execution: %s", requested)
}
